### PR TITLE
KAFKA-3613: Consolidate TumblingWindows and HoppingWindows into TimeWindows

### PR DESCRIPTION
**Pull request by miguno**
@guozhangwang @enothereska @mjsax : Feel free to take a look.

Beyond the consolidation of windows, I also added further unit tests, merged `equalsTo` and `equals`, and renamed `every()` to `shiftedBy()`.

### CODE REVIEW

**Comment by miguno** [comment]
Three tests are failing, stay tuned.

**Comment by miguno** [comment]
Works now.

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+     * @param hop The hop size of the window, with the requirement that 0 &lt; hop &le; size.
+     *            The hop size's effective time unit is determined by the semantics of the
+     *            topology's configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     * @return a new window definition
+     */
+    public TimeWindows shiftedBy(long hop) {
+        return new TimeWindows(this.name, this.size, hop);
+    }
+
+    // TODO: Document windowsFor()
```
Comment (enothereska) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: @miguno I guess you'll remove this TODO as part of PR? Thanks.

**Comment by enothereska** [comment]
LGTM, tests pass for me (above failure seems to be unrelated).

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java (hunk 1 above)
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: @enothereska: Docstring TODO completed.

**Comment by miguno** [comment]
Closed in favor of https://github.com/apache/kafka/pull/1277 (which rebases the changes in this PR on latest Kafka trunk).

**Pull request closed without merging**